### ISSUE TITLE
Add Bioformats Reader

### ISSUE THREAD

**Issue by tischi** [bigdataprocessor/bigdataprocessor2#124]
@nicokiaru
Are you actually interested in doing this in order to get a bit in touch with the code?
I can also do it, no problem...

**Comment by NicoKiaru** [comment]
Sure! But are you a rush ? I'd like to try on Monday morning, if that's ok for you.

**Comment by tischi** [comment]
We should send the revised paper back end of the month (@nornil, right?)
So Monday is fine!
Great!

**Comment by nornil** [comment]
Thanks Nico! 
Much appreciated.

**Comment by NicoKiaru** [comment]
I had a look, and finally it seems more complex than expected to include the bioformats reader, because the `CachedCellImage` relies on specific classes of this project. `FileInfos` and `Image` are only seen here, and transforming the more standard bdv class into these structures is more involved than I expected.

You can check the branch `include_bioformats` and tell me what you think, and see where I'm stuck.

**Comment by NicoKiaru** [comment]
It's also related to these issues:
https://github.com/bigdataprocessor/bigdataprocessor2/issues/2
https://github.com/bigdataprocessor/bigdataprocessor2/issues/58

Can bigdataprocessor be changed so that it can use other image loader than [`ImageLoader`](https://github.com/bigdataprocessor/bigdataprocessor2/blob/master/src/main/java/de/embl/cba/bdp2/open/core/ImageLoader.java)?

**Comment by tischi** [comment]
> Can bigdataprocessor be changed so that it can use other image loader than ImageLoader?

Absolutely. The `ImageLoader` is only used by the `CachedCellImgCreator`, which are both very specific to the import of custom Tiff or Hdf5 based image sequences. I think I will rename both classes to make this more clear.

What you could try to think about is simply whether you would manage to create an `Image` where the internal `raiXYZCT` is backed by a `CachedCellImg` that loads the data through your bioformats library. If this would work, I think this would be the easiest, because I am afraid a lot of my code depends on the `raiXYZCT` and it would be quite some work to change this now in the timeline of the review.

**Comment by NicoKiaru** [comment]
Yes, but basically I'm stuck at this point : from the CachedCellImg to a raiXYCZT.

If you look at my branch here:

https://github.com/bigdataprocessor/bigdataprocessor2/blob/bf1901acb1685413357fe943d0fa13f0d2f1b801/src/main/java/de/embl/cba/bdp2/open/core/CachedCellImgCreator.java#L65

The rai is a cached cell img (only XYZ, but they can be stacked with views into a raiCZT). But it looks like `FileInfos` is needed as well. Does it ? If yes, I don't know how to instanciate it.

**Comment by tischi** [comment]
Don't use my `CachedCellImgCreator`!
Just make your own one.
I will rename mine to something like `FileSeriesCachedXYZCTCellImgCreator`.
Yours can be `BioFormatsCachedXYZCTCellImgCreator`.

**Comment by tischi** [comment]
...and the `FileInfos` are only necessary for the `FileSeriesCachedXYZCTCellImgCreator`.

**Comment by tischi** [comment]
I will refactor everything to make it more clear.
I hope you can still keep your code when you merge the new master...

**Comment by tischi** [comment]
I found out why I need the FileInfos again later, but this is not good and I have an idea how to fix it.

The point is the following:

- For fast browsing through the image along the z-axis it is best if the backing `CachedCellImg` is chunked plane-wise in terms of its cells.
- For fast processing of the whole data set it is however better if the backing `CachedCellImg` is chunked volume-wise (one cell per channel and time point) in terms of its cells, because the data can potentially be loaded much faster this way.

Thus, I created a function that can replace the backing `CachedCellImg` on demand: see `CachedCellImgReplacer`.

However, to do this I need to create the volume-wise `CachedCellImg` at some point.

My proposal is now: the first moment the Image is created we store two versions of the `CachedCellImg` inside the `Image` such that one can later ask the `Image` itself via some method, e.g. `switchCacheMode( CacheMode planeWise or volumeWise )`, to do the switch.

Question to you: using your BF library can you do this? That is, create two different versions with different cell sizes?

**Comment by NicoKiaru** [comment]
Yes you can set the cache size or let bioformats decide (it's an option in the builder). Anyways bioformats read plane by plane. So on the first access it will be plane by plane. But if you need to access the volume several times, then maybe the caching by 'volume' block could be useful. So only worth if you need to access several times the same block.

**Comment by tischi** [comment]
>  Anyways bioformats read plane by plane.

Ok, so we cannot force it to read a whole volume in one go?

> So only worth if you need to access several times the same block.

Ok, then it is not so useful, because I read only once in the scenario where my code needs to read the whole volume.

------

Anyway, I thought of another design idea:

1. We make a `CachedCellImgCreator` an interface, with implementations:
  - FileSeries (done by me)
  - BioFormats (done by you)
2. We add a `CachedCellImgCreator` private field inside `Image`
3. We add a method in `Image` that allows to `setCachingMode( int[] cellSizeXYZCT, blocked/softref, int cacheSize )`
  - This method would call the `cachedCellImgCreator` to create the new desired `cachedCellImg` and the use the `CachedCellImgReplacer` class to change the caching of the `raiXYZCT`.
  - If then in the BioFormats case this will happen to not increase read speed, that's just life. It would still work.

Essentially `Image` then is a class that represents an `raiXYZCT`, with a `voxelSize` and enables flexible configuration of the backing of the `raiXYZCT`, which is needed because the `raiXYZCT` can be TB-sized.
I think that seems sensible. What do you think?

**Comment by NicoKiaru** [comment]
> Ok, so we cannot force it to read a whole volume in one go?

The 'core function' works by plane : 
https://github.com/BIOP/bigdataviewer-bioformats/blob/1dc74f5853088b8b43b608d087e5b7927672727f/src/main/java/ch/epfl/biop/bdv/bioformats/bioformatssource/BioFormatsBdvSourceUnsignedByte.java#L106

However in `bigdataviewer-bioformats`, you can make the cell loader read a bunch of plane during 'one' cell loading event:
https://github.com/BIOP/bigdataviewer-bioformats/blob/1dc74f5853088b8b43b608d087e5b7927672727f/src/main/java/ch/epfl/biop/bdv/bioformats/bioformatssource/BioFormatsBdvSourceUnsignedByte.java#L85

This just avoids several calls of the CellLoader, and the reader is already available for different planes. So maybe it's a tiny bit faster if the cache is set to volume. But don't expect magic.

Im a bit lost with your explanation right now but if you can give it a go I'll check on monday. My goal this monday will first be to create a raixyzct from `bigdataviewer-bioformats`. The caching can be tuned when the source created (cache size, softref and so on), but not after the CachedCellImg are made.

**Comment by tischi** [comment]
> Im a bit lost with your explanation right now 

I understand, fair enough!

> My goal this monday will first be to create a raixyzct from bigdataviewer-bioformats. The caching can be tuned when the source created (cache size, softref and so on), but not after the CachedCellImg are made.

Perfect! That's all we need. I can probably do the rest.

**Comment by tischi** [comment]
@NicoKiaru 
I may not manage to refactor all the code until Monday. 
However, if you could work on implementing below interface that would be very valuable for sure!

```Java
public interface CachedCellImgCreator < R extends RealType< R > & NativeType< R > >
{
	String getImageName();

	String[] getChannelNames();
	
	double[] getVoxelSize();
	
	String getVoxelUnit(); // TODO: maybe rather something with a controlled vocabulary here?  For example is there an enum in BioFormats?

	CachedCellImg< R, ? > createCachedCellImg( int[] cellDimsXYZCT, DiskCachedCellImgOptions.CacheType cacheType, long cacheSize );
	
	// TODO: maybe add something like getDisplaySettings();
}
```

**Comment by NicoKiaru** [comment]
I did a stacking into raixyzct and I checked that it works with czi (time lapse and z stacks).

It's in the include-bioformats branch. However I increased the version of bdv-core and imglib2-cache.

This broke a few things...

But I now have a working raixyzct, so let m eknow when the rest is done and we can try to merge these things:

https://github.com/bigdataprocessor/bigdataprocessor2/blob/06d74bb70c234ca014c20085c08b45afa1f9374d/src/main/java/de/embl/cba/bdp2/open/core/CachedCellImgCreator.java#L40

**Comment by tischi** [comment]
New interface

```Java
public interface CachedCellImgCreator< R extends RealType< R > & NativeType< R > >
{
	String getImageName();

	String[] getChannelNames();

	ARGBType[] getChannelColors();
	
	double[] getVoxelSize();

	Unit< Length > getVoxelUnit();
	
	int[] getDefaultCellDimsXYZCT(); // "default": good for fast browsing in BDV

	CachedCellImg< R, ? > createCachedCellImg( int[] cellDimsXYZCT, DiskCachedCellImgOptions.CacheType cacheType, long cacheSize );
}
```

**Comment by tischi** [comment]
I am using this class from you:

```
this.voxelUnit = BioFormatsMetaDataHelper.getUnitFromString( fileInfos.voxelUnit );
```

You think that's ok?

However, it does not work for the `getSymbol()` case, but returns NULL:

![image](https://user-images.githubusercontent.com/2157566/95769791-b67b9400-0cb8-11eb-996b-6e581d3f4f95.png)

Do you know how to fix this? Maybe BioFormats itself has a function for this?

**Comment by tischi** [comment]
I pushed changes to `master`, so you could work from there.
I already added a dummy `BioFormatsCachedCellImgCreator` class for you to start with.

**Comment by NicoKiaru** [comment]
> I pushed changes to `master`, so you could work from there.
> I already added a dummy `BioFormatsCachedCellImgCreator` class for you to start with.

Great!

> However, it does not work for the `getSymbol()` case, but returns NULL:

I forgot about this function... Its implementation is a bit  weird, I don't remember why. https://github.com/BIOP/bigdataviewer-bioformats/blob/1dc74f5853088b8b43b608d087e5b7927672727f/src/main/java/ch/epfl/biop/bdv/bioformats/BioFormatsMetaDataHelper.java#L672

**Comment by tischi** [comment]
...yeah but it does _not_ work for `µm`. Could we make it also work for this?

**Comment by NicoKiaru** [comment]
Maybe try to add the method `getSymbol` in the test:

```
if ((f.getName().toUpperCase().equals(unit_string.trim().toUpperCase())||(f.getSymbol().toUpperCase().equals(unit_string.trim().toUpperCase())) {
```

https://github.com/BIOP/bigdataviewer-bioformats/blob/1dc74f5853088b8b43b608d087e5b7927672727f/src/main/java/ch/epfl/biop/bdv/bioformats/BioFormatsMetaDataHelper.java#L677

You can take the whole function

**Comment by NicoKiaru** [comment]
Maybe this:

```
    public static Unit getUnitFromString(String unit_string) {
        Field[] bfUnits = UNITS.class.getFields();
        for (Field f:bfUnits) {
            if (f.getType().equals(Unit.class)) {
                if (f.getName()!=null) {
                     try {
                         if ((f.getName().toUpperCase().equals(unit_string.trim().toUpperCase())||(((Unit)(f.get(null))).getSymbol().toUpperCase().equals(unit_string.trim().toUpperCase())){//(f.getName().toUpperCase().equals(unit_string.trim().toUpperCase())) {
                             // Field found
                            Unit u = (Unit) f.get(null); // Field is assumed to be static
                            return u;
                         }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        }
        // Field not found
        return null;
    }
```

**Comment by tischi** [comment]
Thanks! That works! 
Could you update it in `bdv-bdf`?
I would immediately pull the changes because I need this to work.

**Comment by NicoKiaru** [comment]
Ah, I had to modify this bdv-bdf back because of some issue with the bdv-playground dependency removal.

But this a static self-contained class, can't you copy it for the moment ? Sorry but I won't have time today to fix it

**Comment by tischi** [comment]
No worries, I can wait, in fact.
I rather not copy it to avoid code duplication.
I can work on other stuff....

**Comment by NicoKiaru** [comment]
Hi @tischi , 

please have a look at https://github.com/bigdataprocessor/bigdataprocessor2/tree/include_bioformats

I have an issue : I do work with `CachedCellImg`, but only 3d. When I stack the views to stick to a 5D RAI, then the RAI cannot be cast to a `CachedCellImg` anymore. Is there a way around this ?

**Comment by tischi** [comment]
> I have an issue : I do work with CachedCellImg, but only 3d. When I stack the views to stick to a 5D RAI, then the RAI cannot be cast to a CachedCellImg anymore. Is there a way around this ?

I think the easiest may be instead of the `Views.stack` rather use a `CachedCellImg` (via its loader) to combine the data (so you would construct a 5D loader from the 3D data cubes (which you could store as lists within the loader).

**Comment by tischi** [comment]
...in terms of performance not ideal because the code will copy the data twice, but we can worry about the later. I would just think it is a good idea to set the cache size of your initial 3D cachedCellImgs to `0` such that we do not cache the data two times.

**Comment by NicoKiaru** [comment]
Do you have a good reason not to return a RAI in `createCachedCellImg` ?

**Comment by tischi** [comment]
Yes, I think so. If you look at the `CachedCellImgReplacer` class, which is called here in my `Image` class:

```
	public void setCache( int[] cellDims, DiskCachedCellImgOptions.CacheType cacheType, int cacheSize )
	{
		Logger.info( "Setting cache of image: " + getName() );
		Logger.info( "  Cache cell dimensions: " + Arrays.toString( cellDims ) );
		Logger.info( "  Cache size: " + cacheSize );
		Logger.info( "  Cache type: " + cacheType.toString() );

		this.cellDims = cellDims;
		CachedCellImg< R, ? > cachedCellImg = cachedCellImgCreator.createCachedCellImg( cellDims, cacheType, cacheSize );

		if ( rai == null )
			rai = cachedCellImg;
		else
			rai = new CachedCellImgReplacer<>( rai, cachedCellImg ).get();
	}
```

This call requires a `cachedCellImg`:
```
new CachedCellImgReplacer<>( rai, cachedCellImg ).get();
```

And I also checked my code inside `CachedCellImgReplacer` and there this also seems necessary.

But, the `CachedCellImgReplacer` currently is only needed for performance reasons.
So, you could do the following:

1. change the interface to return an `RAI` (make sure you pulled the latest master)
1. in the `setCache` function of the `Image` class do an `instanceof CachedCellImg` and if it is not `instanceof CachedCellImg` the setCache would do nothing in case an `RAI` exists already
1. make an issue that your bioformats reader may be changed in the future to return a `CachedCellImg`

**Comment by NicoKiaru** [comment]
>     3\. make an issue that your bioformats reader may be changed in the future to return a `CachedCellImg`

In fact the bioformats reader returns a `CachedCellImg`, but one per timepoint and per channel. I'm then using `Views.stack` with these CachedCellImgs to make a 5D RAI. So in fact it's `Views.stack` which do not preserve the caching structure. I looked at `Views.stack` to see if it was possible to make a `CachedViews.stack` but that looks tedious.

I'll check what can be done but not sure if I have time this week.

**Comment by tischi** [comment]
> I'll check what can be done but not sure if I have time this week.

As I suggested, maybe leave this for later (see my points 1 and 2 above).
Because then we can work already on the UI before fixing the issue.

Ok?

**Comment by tischi** [comment]
> Views.stack which do not preserve the caching structure. I looked at Views.stack to see if it was possible to make a CachedViews.stack but that looks tedious.

I don't think that's the way to go, because Views put another layer on top of the Cache. I think this is the design of imglib2, I would not change this. I think you we should directly create a 5D Cache using a CellLoader.

Anyway, as said above, for now you can just return an RAI. If you don't have much time maybe just paste here your current CachedCellImgCreator class and I integrate it.

**Comment by NicoKiaru** [comment]
Ok, it works now in the branch include_bioformats.

Take care if you want to merge with master because I had to bump the version of bigdataviewer-core and vistools. So it will not working with other update sites such as BigStitcher.

Colors and pixel size are ok. I changed the factory to return a RAI, this did not require a lot of modification, but you can check whether this causes a problem in downstream processing (which I did not test).

**Comment by tischi** [comment]
OK!
Do we really need the new bdv-core and vistools?
bdv-bioformats should not even need those dependencies, should it?

**Comment by NicoKiaru** [comment]
It may be possible to avoid it. Do not merge with master then, I can try to do another branch, now that I see what can be done.

**Comment by tischi** [comment]
ok, I think this would be great (also for your code ;-)

**Comment by NicoKiaru** [comment]
The issue is this `DisplaySettings` property stuff. I think I'll make a stupid independent repo containing only this just to be able to split bdv-playground and bdv-bioformats.

**Comment by tischi** [comment]
...or you put it into `bdv-bioformats` in case you always want to have bdv-bioformats as a dependency within playground.
I think that way around it could make sense, because the playground is some kind of melting pot anyway.

**Comment by NicoKiaru** [comment]
Ok, the bioformats reader is done ad works in the branch include_bf_new  . Please check it

**Comment by tischi** [comment]
Awesome! Works! See new master!

I think the next issue would be how we let the user choose the `series`.
I think a starting point could be a class like

```
class BioFormatsInspector
{
    BioFormatsInspector( String filePath ) { } 
    public getNumSeries( String filePath )
}
```

For example if `getNumSeries( String filePath )` returns `1` we do not show a UI for it. 
For resubmission we could even throw an error for multi-series files and say that we will add support later, on demand.

...we would then also remove the `series` from the `Command` and work more with `GenericDialogs` wizard style, only showing the choices if they make sense. 

What do you think?

**Comment by NicoKiaru** [comment]
Maybe re-use what there is in the standard bioformats gui ? I don't know where this is

**Comment by tischi** [comment]
> Maybe re-use what there is in the standard bioformats gui ? I don't know where this is

Long term: yes; Short term: too much work / not enough time.

**Comment by tischi** [comment]
ok, for now I just did like this:
```
            BioFormatsInspector inspector = new BioFormatsInspector( filePath );
            if ( inspector.getSeriesCount() > 1 )
            {
                Logger.warning( "File contains multiple image series, opening the first one." );
                Logger.warning( "If you need better support for multi-series files please report here:" );
                Logger.warning( "https://forum.image.sc/t/bigdataprocessor/34963" );
            }
```

**Comment by NicoKiaru** [comment]
But right now the user can choose the series number (it's a parameter in the command). Did you remove this option ?

**Comment by tischi** [comment]
> Did you remove this option ?

Yes, because if they do not know how many there are and which is which it does not really help.

**Comment by tischi** [comment]
@NicoKiaru 
In your next release of bdv-bf, will this also include the update of this function?
```
this.voxelUnit = BioFormatsMetaDataHelper.getUnitFromString( fileInfos.voxelUnit );
```
...I remember there was some issue that you fixed.

**Comment by NicoKiaru** [comment]
Wasn't it already done ?

**Comment by tischi** [comment]
I think so, I was just not sure in _which_ _version_ of bdv-bf.
